### PR TITLE
Tweak locations

### PR DESCRIPTION
**Pull request by Drup**
This is a small patch to slightly tweak locations in js files. There are two objectives:
- Add more locations in the JS AST, notably for variables, which allow to easily look them up in srcmaps
- Try to Add more srcmap infos on linked js files.

We could probably do more on the topic, but it's sufficient for the current side-project I'm working on (which I'll show soon :p)

### CODE REVIEW

**Comment by Drup** [comment]
A few things I noticed:
- sourcemaps seems mostly empty when `--pretty` is *not* enabled. Is that intended ?
- Inlining and lifting seems to completely forget about source locations
- There seems to be something wrong with functors. Notably the function representing `Map.Make` lost all his source information

I wouldn't mind some pointers to try to approach those, if you don't feel like tackling them yourself. The jsoo compiler codebase is a bit intimidating. :p

**Comment by hhugo** [comment]
https://github.com/ocsigen/js_of_ocaml/issues/799 should help with testing

**Comment by Drup** [comment]
Do we really need to wait for that in order to merge ? It seems pretty independent to me.

File: compiler/lib/javascript.ml [hunk 1]
```diff
@@ -129,7 +129,10 @@ type identifier = string
 
 type ident_string =
   { name : identifier
-  ; var : Code.Var.t option }
+  ; var : Code.Var.t option
+  ; loc : location}
+
+let mk ?(loc=N) ?var name = { name ; var ; loc }
```
Comment (hhugo) [inline compiler/lib/javascript.ml]: The name `mk` looks a bit weird to me.
What about having a function `ident` instead, defined as
```
let ident ?(loc=N) ?var name = S { name ; var ; loc }
```

**Comment by hhugo** [comment]
I agree, #799 does't have to block this PR

File: compiler/lib/javascript.ml (hunk 1 above)
Comment (Drup) [reply compiler/lib/javascript.ml]: Fixed.

**Comment by hhugo** [comment]
I took the liberty to add 2 commits to your PR.
I'll merge once you confirm you're happy with these changes and CI is green.

File: compiler/lib/js_output.ml [hunk 2]
```diff
@@ -127,10 +127,13 @@ struct
       | Some _ -> ()
 
   let ident f = function
-    | S {name; var = None} -> PP.string f name
-    | S {name; var = Some v} ->
-        output_debug_info_ident f name v;
+    | S {name; var = Some v; _} ->
```
Comment (hhugo) [inline compiler/lib/js_output.ml]: Here, we ignore `loc` and only try to extract a loc from `var`. I wonder if we should have a pass that try populate all the locs

Comment (Drup) [reply compiler/lib/js_output.ml]: Can sourcemap accomodate multiple bindings for one identifier ?

I think a better longterm solution is to enforce (by datatype) that there is a loc in either `v` or `pi`.

**Comment by Drup** [comment]
Looks good to me!

**Pull request merged**